### PR TITLE
Add --user-agent daemon flag

### PR DESCRIPTION
**Pull request by jonjohnsonjr**
**\- What I did**

Added `--user-agent` flag to daemon that adds an annotation to the User-Agent header.

For example, the flag `--user-agent "my-annotation"` yields this user-agent:

`docker/1.13.0-dev go/go1.7.1 git-commit/c2a6ba7 kernel/3.13.0-95-generic os/linux arch/amd64 custom/my-annotation UpstreamClient(Docker-Client/1.13.0-dev \\(linux\\))`

Note the new "custom/my-annotation". I'm open to suggestions on a better name than "custom" or a nice method for specifying the [product](https://tools.ietf.org/html/rfc7231#section-5.5.3). :)

Why is this useful?

We have multiple runtime environments using docker, but no simple way to distinguish between these environments and normal docker users (for analytics and debugging purposes). This flag solves that.

Are you aware of the [HttpHeaders](https://docs.docker.com/engine/reference/commandline/cli/#/configuration-files) config property?

Yes. This isn't a great solution for us, since it requires modifying the client configuration file. AFAICT, there's no programmatic way to manipulate the client config file, so this isn't a particularly robust option. It's also much simpler to add this to the daemon than try to plumb it through clients, since we have more control over the daemon configuration.

**\- How I did it**

To make this a simpler change, I broke it into four commits.

The first two move context handling for the user-agent and docker version from `httputils` to `pkg/useragent` and `dockerversion`. I also renamed the functions to `NewContext` and `FromContext`, which [I believe is more idiomatic](https://blog.golang.org/context):
- `getUserAgentFromContext` becomes `useragent.FromContext`
- `httputils.VersionFromContext` becomes `dockerversion.FromContext`

The third commit re-centralizes the user-agent logic to effectively undo parts of #19057. That PR pulled the user-agent logic out of the `registry` package and added the user-agent as a parameter to a bunch of functions. In #21373, a context parameter was added back to most of those functions (or their callers), and an implicit dependency on the user_agent middleware (`api/server/middleware/user_agent.go`) was introduced; i.e. calls to `dockerversion.DockerUserAgent` expect the user_agent middleware to have previously populated ctx with the upstream client's user-agent.

I think it's cleaner to have the user_agent middleware just construct the user-agent and add it to the context instead. This makes `dockerversion.DockerUserAgent`'s dependency on the user_agent middleware explicit and IMO makes it much simpler to reason about how the user-agent is constructed, since it all happens in one place. As a happy side effect, we can remove the user-agent parameter from any functions that take a context.

The final commit actually adds the flag. I'm happy to squash these into one commit or break them apart into multiple PRs -- not sure what the protocol is.

**\- How to verify it**

Run the integration test:

```
TESTFLAGS='-check.f DockerRegistrySuite.TestUserAgentPassThrough' make test-integration-cli
```

Or start dockerd with the flag:

```
dockerd --user-agent "my-annotation" &
```

and check the logs of a push against a local registry:

```
docker run -d -p 5000:5000 registry:2 && \
 docker pull busybox && \
 docker tag busybox localhost:5000/busybox && \
 docker push localhost:5000/busybox && \
 docker logs `docker ps -q` # assuming no other running containers
```

You should see "custom/my-annotation". 

**\- Description for the changelog**

Add `--user-agent` daemon flag to annotate the user-agent with additional info (for debugging, analytics).

**\- A picture of a cute animal (not mandatory but encouraged)**

Say hi to Toxic!

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
I wonder this should be implemented in the daemon side..
Did you consider implementing the feature in a custom HTTP proxy?

e.g. https://github.com/nodejitsu/node-http-proxy#setup-a-stand-alone-proxy-server-with-proxy-request-header-re-writing

**Comment by thaJeztah** [comment]
I'm -1 in this; this is adding a lot of changes for a very narrow use case. If this is on an internal network, wouldn't you be able to just analyze based on IP address?

**Comment by mattmoor** [comment]
It's not a very narrow use case, the fundamental problem is the breadth of Docker usage we're trying to track.  We (Google) are a major cloud provider with multiple heavily adopted container runtime offerings and our own Registry that serves all of them.  We would like a way to meaningfully distinguish traffic between the major runtime counterparts.

re: IPs
This simply doesn't work as the different runtime partners we have don't have isolated IP blocks we can filter on, it all essentially looks like Google Compute Engine to us.

re: running a proxy
At the scale we're talking about it would make more sense to simply partition at the registry DNS level, but this is an annoying burden to push onto our customers who want a single registry (per region) that they can use to manage their images.


We simply want to be able to configure the nodes in our runtime clusters with the flavor of container runtime that is orchestrating the pulls we are seeing.  Again, given the enormous number of container orchestration systems out there (Swarm, Mesos, Kubernetes, ECS, ...) this is hardly a niche.

Any guidance on a path forward would certainly be appreciated.

**Comment by thaJeztah** [comment]
@mattmoor @jonjohnsonjr would this approach help for your use case (or a similar approach?) https://github.com/docker/docker/pull/27509

**Comment by runcom** [comment]
I'm +1 on this. The change isn't huge (apart from a refactor) and it's just adding a flag on the daemon to just add an annotation in the user agent sent to a registry...

**Comment by thaJeztah** [comment]
In general, we'd be more open to adding additional headers to the requests instead of modifying the user agent

**Comment by LK4D4** [comment]
@runcom @mattmoor @j0hnst0n so, what do you think about additional headers?

**Comment by jonjohnsonjr** [comment]
Additional headers seems fine. Any suggestions for a name? I'm not super familiar with Docker naming conventions.

## Proposal

1. Keep the first 3 user-agent refactoring commits.
2. Take my changes from [a09a019](https://github.com/docker/docker/pull/26936/commits/a09a019fae91edde772d6b8c7c4e135a4b5d856a) and move them to a new "trace" middleware that does basically the same thing with a different name.
3. Modify DockerHeaders to take a context instead of userAgent string, then pull both "User-Agent" and this trace annotation from the context.

(I think #27509 could then just add "Docker-Session-ID" to the context in the same middleware. Session ID would span one docker command; this annotation would span multiple commands.)

What do you think?

Alternatively, I could add "HttpHeaders" to the daemon config (similar to the client config) and do something similar. That would be a general mechanism for attaching headers to requests, which maps pretty well to the open tracing's baggage. This PR + #27509 could then be refactored to use open tracing.

File: dockerversion/version.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// FromContext returns the API version stored in ctx, if one exists.
+func FromContext(ctx context.Context) (v string) {
+	if v, ok := ctx.Value(APIVersionKey).(string); ok {
+		return v
+	}
+	return
+}
+
+// NewContext returns a new Context that carries the API version.
+func NewContext(ctx context.Context, version string) context.Context {
```
Comment (stevvooe) [inline dockerversion/version.go]: `WithVersion` would be the correct pattern here.

File: dockerversion/version.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package dockerversion
+
+import (
+	"golang.org/x/net/context"
+)
+
+// APIVersionKey is the client's requested API version.
+const APIVersionKey = "api-version"
```
Comment (stevvooe) [inline dockerversion/version.go]: Please follow the best practices here. This should be a local type:

```
type apiVersionKey struct{}
```

This keeps storage requirements small and access local to this package.

**Comment by stevvooe** [comment]
Might be more clear if this flag were `--user-agent-component` to make it clear that this isn't a wholesale replacement of the user agent. It would also be better if we could make this consistent on client and server, if possible.

**Comment by thaJeztah** [comment]
It looks like this PR is currently not in a mergeable state, and we're not fully complete on design. Let me close this PR for now, but feel free to continue the discussion, and let me know if you want to start working on this again @jonjohnsonjr, then I can reopen

**Pull request closed without merging**